compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
         Some(_) | None => {}
     };
 
-    let filter = |s: &str| {
-        if s.is_empty() {
-            return vec![];
-        }
-        let feature = if s.starts_with('+') || s.starts_with('-') {
-            &s[1..]
-        } else {
-            return vec![s.to_string()];
-        };
-        // Rustc-specific feature requests like `+crt-static` or `-crt-static`
-        // are not passed down to LLVM.
-        if RUSTC_SPECIFIC_FEATURES.contains(&feature) {
-            return vec![];
-        }
-        // ... otherwise though we run through `to_llvm_feature` feature when
-        // passing requests down to LLVM. This means that all in-language
-        // features also work on the command line instead of having two
-        // different names when the LLVM name and the Rust name differ.
-        to_llvm_feature(sess, feature).iter().map(|f| format!("{}{}", &s[..1], f)).collect()
-    };
-
     // Features implied by an implicit or explicit `--target`.
-    features.extend(sess.target.features.split(',').flat_map(&filter));
+    features.extend(
+        sess.target
+            .features
+            .split(',')
+            .filter(|v| !v.is_empty() && backend_feature_name(v).is_some())
+            .map(String::from),
+    );
 
     // -Ctarget-features
-    features.extend(sess.opts.cg.target_feature.split(',').flat_map(&filter));
+    let supported_features = supported_target_features(sess);
+    features.extend(
+        sess.opts
+            .cg
+            .target_feature
+            .split(',')
+            .filter_map(|s| {
+                let enable_disable = match s.chars().next() {
+                    None => return None,
+                    Some(c @ '+' | c @ '-') => c,
+                    Some(_) => {
+                        let mut diag = sess.struct_warn(&format!(
+                            "unknown feature specified for `-Ctarget-feature`: `{}`",
+                            s
+                        ));
+                        diag.note("features must begin with a `+` to enable or `-` to disable it");
+                        diag.emit();
```

Comment:
I think this warning should be moved to the driver, so it also works for other codegen backends and for `cargo check`.

Reply:
Given that each backend is currently in charge of parsing everything about the target features, except for splitting by comma. I don't think right now is the right time to move this diagnostic to anywhere.

I do have more changes I want to make with regards to how we handle target features, and moving some of that responsibility away from `rustc_codegen_*` is one of those things. As an example, the only thing codegen backends should probably know about “rustc” target features is how to convert a rustc target feature to its own internal interpretation.

But I think some more refactors are in order before this can happen.